File: src/main/java/training/official/classic_puzzle_easy/Defibrillators.java
package training.official.classic_puzzle_easy;

import java.util.Scanner;

class Defibrillators {
	private static final int NAME = 1;
	private static final int LONGITUDE = 4;
	private static final int LATITUDE = 5;
	private static final int EARTH_RADIUS = 6371;

	public static void main(String args[]) {
		Scanner in = new Scanner(System.in);
		Double userLongitude = convertPositionToDouble(in.next());
		Double userLatitude = convertPositionToDouble(in.next());
		int N = in.nextInt();
		if (in.hasNextLine()) {
			in.nextLine();
		}

		String answer = "";
		Double currentSmallestDistance = Double.MAX_VALUE;

		for (int i = 0; i < N; i++) {
			String[] defib = in.nextLine().split(";");

			Double defibLongitude = convertPositionToDouble(defib[LONGITUDE]);
			Double defibLatitude = convertPositionToDouble(defib[LATITUDE]);

			Double distanceToUser = calculateGpsCoordinatesDistance(userLongitude, userLatitude, defibLongitude, defibLatitude);

			if (distanceToUser < currentSmallestDistance) {
				currentSmallestDistance = distanceToUser;
				answer = defib[NAME];
			}
		}

		System.out.println(answer);
	}

	private static double convertPositionToDouble(String position) {
		return Double.parseDouble(position.replace(',', '.'));
	}

	private static double calculateGpsCoordinatesDistance(Double longitudeA, Double latitudeA, Double longitudeB, Double latitudeB) {
		Double x = (longitudeB - longitudeA) * Math.cos((latitudeA + latitudeB) / 2);
		Double y = latitudeB - latitudeA;
		return Math.sqrt(x * x + y * y) * EARTH_RADIUS;
	}
}

File: src/main/kotlin/training/official/classic_puzzle_easy/Defibrillators.kt
//package training.official.classic_puzzle_easy

import java.util.*
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sqrt

fun main() {
	val (position, defibrillators) = Defibrillators.extractInputData()

	val closestDefibrillator = defibrillators.minByOrNull { defibrillator -> Defibrillators.calculateDistance(position, defibrillator.position) }

	println(closestDefibrillator?.name ?: "")
}

private class Defibrillators { // only here for codingame
	companion object {
		fun calculateDistance(userPosition: Input.Position, defibrillatorPosition: Input.Position): Double {
			return calculateDistance(userPosition.longitude, userPosition.latitude, defibrillatorPosition.longitude, defibrillatorPosition.latitude)
		}

		fun calculateDistance(longitudeA: Double, latitudeA: Double, longitudeB: Double, latitudeB: Double): Double {
			val earthRadius = 6371
			val x = (longitudeB - longitudeA) * cos((latitudeA + latitudeB).div(2))
			val y = (latitudeB - latitudeA)
			return sqrt(x.pow(2) + y.pow(2)) * earthRadius
		}

		fun extractInputData(): Input {
			return Scanner(System.`in`).run {
				val longitude = next()
				val latitude = next()
				val numberOfDefibrillators = nextInt()
				if (hasNextLine()) {
					nextLine()
				}
				val defibrillators = (0 until numberOfDefibrillators)
					.map { nextLine() }
					.map { it.split(";") }
					.map { Input.Defibrillator(it[1], Input.Position(it[4], it[5])) }
					.toList()
				Input(Input.Position(longitude, latitude), defibrillators)
			}
		}
	}

	data class Input(val position: Position, val defibrillators: List<Defibrillator>) {
		data class Position(private val _longitude: String, private val _latitude: String) {
			val longitude: Double
				get() = _longitude.replaceCommaWithDot().toDouble()
			val latitude: Double
				get() = _latitude.replaceCommaWithDot().toDouble()

			private fun String.replaceCommaWithDot() = this.replace(",", ".")
		}

		data class Defibrillator(val name: String, val position: Position)
	}
}
